packages/gatsby/src/redux/__tests__/index.js

```diff
@@ -149,4 +173,178 @@ describe(`redux db`, () => {
 
     expect(mockWrittenContent.has(legacyLocation)).toBe(false)
   })
+
+  describe(`Sharding`, () => {
+    afterAll(() => {
+      v8Serialize.mockRestore()
+      v8Deserialize.mockRestore()
+    })
+
+    // we set limit to 1.5 * 1024 * 1024 * 1024 per shard
+    // simulating size for page and nodes will allow us to see if we create expected amount of shards
+    // and that we stitch them back together correctly
+    const nodeShardsScenarios = [
+      {
+        numberOfNodes: 50,
+        simulatedNodeObjectSize: 5 * 1024 * 1024,
+        expectedNumberOfNodeShards: 1,
+      },
+      {
+        numberOfNodes: 5,
+        simulatedNodeObjectSize: 0.6 * 1024 * 1024 * 1024,
+        expectedNumberOfNodeShards: 3,
+      },
+    ]
+    const pageShardsScenarios = [
+      {
+        numberOfPages: 50,
+        simulatedPageObjectSize: 10 * 1024 * 1024,
+        expectedNumberOfPageShards: 1,
+      },
+      {
+        numberOfPages: 5,
+        simulatedPageObjectSize: 0.9 * 1024 * 1024 * 1024,
+        expectedNumberOfPageShards: 5,
+      },
+    ]
```

Comment:
This is to construct scenario matrix - and we check each combination. I didn't go over the board - primary purpose of this is that sharding for pages and nodes is separate and that one doesn't affect the other.

Something to think about is wether we want to add some edge cases here - like when number of nodes or pages is equal 0 - for now implementation for those is that we discard persisted state when either is 0 - it does make a whole lot of sense for nodes as core creates some nodes of their own even if user doesn't use any source plugins. It's a bit different for pages - gatsby doesn't create pages (well except for `gatsby develop` creating `dev-404-page` ...). I am not sure if it's worth handling this case more gracefully? Do we want to optimize for scenarios where there are 0 pages?